Allow to pass custom fetch for fetching pre-rendered errors (#13403)
withastro/astro: .changeset/big-hats-train.md

---
'astro': minor
---

> Adds a new optional `prerenderedErrorPageFetch` property to `RenderOptions` in the Adapter API to allow adapters to provide custom implementations for fetching prerendered error pages.
> Now, adapters can override the default `fetch()` behavior, for example when `fetch()` is unavailable or when you cannot call the server from itself.
> The following example provides a custom fetch for `500.html` and `404.html`, reading them from disk instead of performing an HTTP call:

Reviewer: ```suggestion
Adds a new optional `prerenderedErrorPageFetch` option in the Adapter API to allow adapters to provide custom implementations for fetching prerendered error pages.

Now, adapters can override the default `fetch()` behavior, for example when `fetch()` is unavailable or when you cannot call the server from itself.

The following example provides a custom fetch for `500.html` and `404.html`, reading them from disk instead of performing an HTTP call:
```
just breaking into paragraphs, and shortening the intro (maybe was too technical?)

> ```js "preRenderedFetch"
> return app.render(request, {
>   preRenderedFetch: async (url: string): Promise<Response> => {

Reviewer: ``````suggestion
```js "prerenderedErrorPageFetch"
return app.render(request, {
  prerenderedErrorPageFetch: async (url: string): Promise<Response> => {
``````
just updating the name in the code sample!

    if (url.includes("/500")) {
        const content = await fs.promises.readFile("500.html", "utf-8");
        return new Response(content, {
          status: 500,
          headers: { "Content-Type": "text/html" },
        });
    }
    const content = await fs.promises.readFile("404.html", "utf-8");
      return new Response(content, {
        status: 404,
        headers: { "Content-Type": "text/html" },
      });
});
```

> If no value is provided, Astro will fallback to its default behavior for fetching error pages.
> Read more about this feature in the [Adapter API reference](/en/reference/adapter-reference/#prerenderederrorpagefetch).

Reviewer: ```suggestion
If no value is provided, Astro will fallback to its default behavior for fetching error pages.

Read more about this feature in the [Adapter API reference](/en/reference/adapter-reference/#prerenderederrorpagefetch).
```
paragraphs

---
'astro': minor
---

Adds a new optional `prerenderedErrorPageFetch` option in the Adapter API to allow adapters to provide custom implementations for fetching prerendered error pages.

Now, adapters can override the default `fetch()` behavior, for example when `fetch()` is unavailable or when you cannot call the server from itself.

The following example provides a custom fetch for `500.html` and `404.html`, reading them from disk instead of performing an HTTP call:
```js "prerenderedErrorPageFetch"
return app.render(request, {
  prerenderedErrorPageFetch: async (url: string): Promise<Response> => {
    if (url.includes("/500")) {
        const content = await fs.promises.readFile("500.html", "utf-8");
        return new Response(content, {
          status: 500,
          headers: { "Content-Type": "text/html" },
        });
    }
    const content = await fs.promises.readFile("404.html", "utf-8");
      return new Response(content, {
        status: 404,
        headers: { "Content-Type": "text/html" },
      });
});
```
If no value is provided, Astro will fallback to its default behavior for fetching error pages.

> Read more about this feature in the [Adapter API reference](/en/reference/adapter-reference/#prerenderederrorpagefetch).

Reviewer 2: ```suggestion
Read more about this feature in the [Adapter API reference](https://docs.astro.build/en/reference/adapter-reference/#prerenderederrorpagefetch).
```
This needs to be absolute, as it's not on the docs site

---

> 'astro': patch

Reviewer 2: ```suggestion
'astro': minor
```

This adds a feature, so is a minor

---

> Allow to pass custom fetch for fetching pre-rendered errors

Reviewer 2: ```suggestion
Allow adapters to pass custom fetch for fetching pre-rendered error pages
```

> Allow adapaters to pass custom fetch for fetching pre-rendered error pages

Reviewer: Just a quick note that I will add some suggestions here for the changeset once the docs are finished and we know exactly what we're saying about the feature!

But, it should start with something of this format:

```suggestion
Adds a new `preRenderedFetch` property to `RenderOptions` in the Adapter API to allow custom fetching of prerendered error pages
```
that 
- mentions the new property by name
- states which API (in this case, the Adapter API) it belongs to
- tells what it's for

And then since this is a minor release and a new feature, we would probably provide an example or explain how it can be used and why it's cool!!

Question about the name though:

When I am looking at the docs PR, the name `preRenderedFetch` doesn't really tell me that this is for **error pages** specifically. I will need the integration devs to look at this and say whether the name is descriptive enough for what it does, or whether there's any potential for confusion. 

For example, it's being added to docs here with these other things that feel like they describe what they are pretty clearly:

![image](https://github.com/user-attachments/assets/38450c8f-7e97-46dc-ac22-6df4a6213c41)

I am worried that if someone just saw `preRenderedFetch` they might assume this is something more generic? Or not associate it with error pages? But, I don't use this API so I will need someone who is more familiar with these integration APIs to verify whether in context, this seems fine! @florian-lefebvre @Fryuni , do you have thoughts here?

Reviewer 3: Something like `prerenderedErrorPageFetch`? It should at least contain "error" IMO

> Allow adapaters to pass custom fetch for fetching pre-rendered error pages

Reviewer: ``````suggestion
Adds a new optional `preRenderedFetch` property to `RenderOptions` in the Adapter API to allow adapters to provide custom implementations for fetching prerendered error pages.

Now, adapters can override the default `fetch()` behavior, for example when `fetch()` is unavailable or when you cannot call the server from itself.

The following example provides a custom fetch for `500.html` and `404.html`, reading them from disk instead of performing an HTTP call:

```js "preRenderedFetch"
return app.render(request, {
  preRenderedFetch: async (url: string): Promise<Response> => {
    if (url.includes("/500")) {
        const content = await fs.promises.readFile("500.html", "utf-8");
        return new Response(content, {
          status: 500,
          headers: { "Content-Type": "text/html" },
        });
    }

    const content = await fs.promises.readFile("404.html", "utf-8");
      return new Response(content, {
        status: 404,
        headers: { "Content-Type": "text/html" },
      });
});
```
If no value is provided, Astro will fallback to its default behavior for fetching error pages.

Read more about this feature in the [Adapter API reference](/en/reference/adapter-reference/#prerenderedfetch).
``````
OK, so now that we have docs fully written, here's an example idea for the changeset that you can take from/start with!

Note, I do think we'll want to change the actual name to include something about error pages. Just update based on whatever happens with that decision.

Reviewer 4: Suggestion applied manually, preview review again
